File: Java/app/src/main/java/org/wgt/widget/demo/MainActivity.java
package org.wgt.widget.demo;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.ViewGroup;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import androidx.activity.EdgeToEdge;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

public class MainActivity extends AppCompatActivity {

    @Nullable
    private WebView mWebView;

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        mWebView = findViewById(R.id.webview);
        WebSettings webSettings = mWebView.getSettings();
        webSettings.setJavaScriptEnabled(true);
        webSettings.setDomStorageEnabled(true);
        mWebView.setWebViewClient(new WidgetWebViewClient());
        mWebView.loadUrl("file:///android_asset/widget.html");
    }

    @Override protected void onPause() {
        if (mWebView!=null){
            mWebView.onPause();
            mWebView.pauseTimers();
        }
        super.onPause();
    }

    @Override protected void onResume() {
        if (mWebView!=null) {
            mWebView.onResume();
            mWebView.resumeTimers();
        }
        super.onResume();
    }

    @Override protected void onDestroy() {
        if (mWebView!=null){
            mWebView.resumeTimers();
            mWebView.loadUrl("about:blank");
            mWebView.stopLoading();
            if (mWebView.getHandler() != null) {
                mWebView.getHandler().removeCallbacksAndMessages(null);
            }
            mWebView.removeAllViews();
            ViewGroup viewGroup;
            if ((viewGroup = ((ViewGroup) mWebView.getParent())) != null) {
                viewGroup.removeView(mWebView);
            }
            mWebView.setWebChromeClient(null);
            mWebView.setTag(null);
            mWebView.clearHistory();
            mWebView.destroy();
            mWebView = null;
        }
        super.onDestroy();
    }

    private class WidgetWebViewClient extends WebViewClient {
        @Override public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            Intent intent = new Intent(MainActivity.this, WidgetActivity.class);
            intent.putExtra("widget_content_url", request.getUrl().toString());
            startActivity(intent);
            return true;
        }
    }

}

File: Kotlin/app/src/main/java/org/wgt/widget/demo/MainActivity.kt
package org.wgt.widget.demo

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.ViewGroup
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity : AppCompatActivity() {


    private var mWebView: WebView? = null


    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        mWebView = findViewById<WebView>(R.id.webview)?.apply {
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true
            webViewClient = WidgetWebViewClient { widgetUrl ->
                val intent = Intent(this@MainActivity, WidgetActivity::class.java)
                intent.putExtra("widget_content_url", widgetUrl)
                startActivity(intent)
            }
            loadUrl("file:///android_asset/widget.html")
        }

    }

    override fun onPause() {
        mWebView?.let {
            it.onPause()
            it.pauseTimers()
        }
        super.onPause()
    }

    override fun onResume() {
        mWebView?.let {
            it.onResume()
            it.resumeTimers()
        }
        super.onResume()
    }

    override fun onDestroy() {
        mWebView?.let {
            it.resumeTimers()
            it.loadUrl("about:blank")
            it.stopLoading()
            it.handler.removeCallbacksAndMessages(null)
            it.removeAllViews()
            (it.parent as? ViewGroup)?.removeView(it)
            it.webChromeClient = null
            it.tag = null
            it.clearHistory()
            it.destroy()
            mWebView = null
        }
        super.onDestroy()
    }

    private class WidgetWebViewClient(
        private val redirectBlock: (String) -> Unit
    ) : WebViewClient() {

        override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
            request?.url.toString().let {
                redirectBlock.invoke(it)
            }
            return true
        }

    }
}